File: Boj-Kotlin&Java/백준10423번전기가부족해.java
package 백준;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.StringTokenizer;

class node implements Comparable<node> {
    int x,y,dist;
    public node(int x, int y, int dist) {
        this.x = x;
        this.y = y;
        this.dist = dist;
    }
    @Override
    public int compareTo(node o) {
        return this.dist-o.dist;
    }
}

public class 백준10423번전기가부족해 {
    static int[] parent;
    static ArrayList<node> list=new ArrayList<>();
    static int[] arr;
    static int n,m,k;
    static StringTokenizer st;
    static int answer=0;
    public static void main(String[] args) throws IOException {
        BufferedReader br=new BufferedReader(new InputStreamReader(System.in));
        String[] input=br.readLine().split(" ");
        n=Integer.parseInt(input[0]);
        m=Integer.parseInt(input[1]);
        k=Integer.parseInt(input[2]);
        arr=new int[k];
        parent=new int[n+1];
        for(int i=1; i<=n; i++)
            parent[i]=i;
        st=new StringTokenizer(br.readLine());
        for(int i=0; i<k; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
            parent[arr[i]]=0;
        }
        for(int i=0; i<m; i++){
            st=new StringTokenizer(br.readLine());
            int x=Integer.parseInt(st.nextToken());
            int y=Integer.parseInt(st.nextToken());
            int dist=Integer.parseInt(st.nextToken());
            list.add(new node(x,y,dist));
        }
        Collections.sort(list);
        for (node tmp : list) {
            if (find(tmp.x) != find(tmp.y)) {
                answer += tmp.dist;
                union(tmp.x, tmp.y);
            }
        }
        System.out.println(answer);
    }
    static int find(int x){
        if(parent[x]==0)
            return 0;
        if(x==parent[x])
            return x;
        return parent[x]=find(parent[x]);
    }
    static void union(int x, int y){
        int px=find(x);
        int py=find(y);
        if(px<py)
            parent[py]=px;
        else
            parent[px]=py;
    }
}

File: Boj-Kotlin&Java/백준 10423번 전기가 부족해.kt
package 백준
private lateinit var parent:IntArray
private val list= arrayListOf<Triple<Int,Int,Int>>()
private lateinit var arr:IntArray
fun main() {
    val (n, m, k) = readLine()!!.split(" ").map { it.toInt() }
    arr= readLine()!!.split(" ").map { it.toInt() }.toIntArray()
    parent=IntArray(n+1){it}
    for(i in arr.indices)
        parent[arr[i]]=0
    repeat(m){
        val (x,y,z)= readLine()!!.split(" ").map { it.toInt() }
        list.add(Triple(x,y,z))
    }
    list.sortWith(compareBy { it.third })
    var answer=0
    for(i in list.indices){
        if(find(list[i].first)!= find(list[i].second)) {
            answer += list[i].third
            union(list[i].first,list[i].second)
        }
    }
    println(answer)
}
private fun find(x:Int):Int{
    if(parent[x]==0)
        return 0
    if(x== parent[x])
        return x
    parent[x]=find(parent[x])
    return parent[x]
}
private fun union(x:Int,y:Int){
    val (px,py)= intArrayOf(find(x), find(y))
    if(px<py)
        parent[py]=px
    else
        parent[px]=py
}